tests/eosio.msig_tests.cpp

```diff
@@ -999,4 +1104,243 @@ BOOST_FIXTURE_TEST_CASE( switch_proposal_and_fail_approve_with_hash, eosio_msig_
    );
 } FC_LOG_AND_RETHROW()
 
+BOOST_FIXTURE_TEST_CASE( check_earliest_exec_time_for_approve_and_unapprove, eosio_msig_tester ) try {
+   static constexpr uint32_t delay_sec = 1000;
+
+   // helper function.
+   static auto calculate_test_exec_time = [&] (uint32_t delay_sec) -> uint32_t {
+      return time_point{control->pending_block_time() + fc::seconds(delay_sec)}.sec_since_epoch();
+   };
+
+   transaction trx = reqauth( N(alice), {permission_level{N(alice), config::active_name}}, abi_serializer_max_time );
+   trx.delay_sec = delay_sec;
+
+   push_action( N(alice), N(propose), mvo()
+                  ("proposer",      "alice")
+                  ("proposal_name", "first")
+                  ("trx",           trx)
+                  ("requested",     vector<permission_level>{{ N(alice), config::active_name }})
+   );
+   // `propose` initially default-constructs `earliest_exec_time` to the default
+   // constructed state of a `std::optional`.
+   BOOST_REQUIRE_EQUAL( fc::optional<time_point>{}.valid(), get_earliest_exec_time(N(alice), N(first)).valid() );
+
+   push_action( N(alice), N(approve), mvo()
+                  ("proposer",      "alice")
+                  ("proposal_name", "first")
+                  ("level",         permission_level{ N(alice), config::active_name })
+   );
+   // `approve` then initializes `earliest_exec_time` to it's specified value
+   // (calculated by the current time point and what was specified in the
+   // `delay_sec` field of the `transaction_header` in the initial proposal), if
+   // the specified authorizations are met (in this case they are).
+   BOOST_REQUIRE_EQUAL( calculate_test_exec_time(delay_sec), get_earliest_exec_time(N(alice), N(first))->sec_since_epoch() );
+
+   transaction_trace_ptr trx_trace;
+   trx_trace = push_action( N(alice), N(unapprove), mvo()
+                             ("proposer",      "alice")
+                             ("proposal_name", "first")
+                             ("level",         permission_level{ N(alice), config::active_name })
+   );
+   // `unapprove` will revert `earliest_exec_time` back to the default
+   // constructed state of a `std::optional` if the specified authorizations are
+   // not met.
+   BOOST_REQUIRE_EQUAL( fc::optional<time_point>{}.valid(), get_earliest_exec_time(N(alice), N(first)).valid() );
+
+   BOOST_REQUIRE( bool(trx_trace) );
```

Comment:
The checks of the trace of the unapprove transaction don't really provide any value for this test case.